### PR TITLE
HOSTEDCP-919: Add AWS cloud controller manager

### PR DESCRIPTION
**Pull request by csrwng**
**What this PR does / why we need it**:
Introduces a new deployment for the AWS cloud controller manager. This is required because in 4.14 kubelets are initialized with an external cloud manager.

Updates the kube cloud controller manager AWS role with additional permissions.

**Checklist**
- [x] Subject and description added to both, commit and PR.

### CODE REVIEW

File: control-plane-operator/controllers/hostedcontrolplane/kcm/params.go [hunk 1]
```diff
@@ -103,10 +102,6 @@ func NewKubeControllerManagerParams(ctx context.Context, hcp *hyperv1.HostedCont
 	params.DeploymentConfig.SetRestartAnnotation(hcp.ObjectMeta)
 
 	switch hcp.Spec.Platform.Type {
-	case hyperv1.AWSPlatform:
-		params.CloudProvider = aws.Provider
```
Comment (muraee) [inline control-plane-operator/controllers/hostedcontrolplane/kcm/params.go:107]: shouldn't we be setting `params.CloudProvider` to `external`?

Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/kcm/params.go:107]: my understanding is that is already getting set on the kubelet by the mco when we generate an ignition payload

Comment (muraee) [reply control-plane-operator/controllers/hostedcontrolplane/kcm/params.go:107]: right, but according to the [docs](https://cloud-provider-aws.sigs.k8s.io/getting_started/) we should set it on the kube-controller-manager as well as on the kubelet.

Comment (csrwng) [reply control-plane-operator/controllers/hostedcontrolplane/kcm/params.go:107]: ahh, ic what you mean, thx
If the current run passes, I'd rather merge it so we can get CI unstuck and I can update in a followup.

File: control-plane-operator/controllers/hostedcontrolplane/cloud/aws/reconcile.go [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+		c.Image = controllerManagerImage
+		c.ImagePullPolicy = corev1.PullIfNotPresent
+		c.Command = []string{"/bin/aws-cloud-controller-manager"}
+		c.Args = []string{
+			"--cloud-provider=aws",
+			"--use-service-account-credentials=false",
+			"--kubeconfig=/etc/kubernetes/kubeconfig/kubeconfig",
+			"--cloud-config=/etc/cloud/aws.conf",
+			"--configure-cloud-routes=false",
+			"--leader-elect=true",
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/cloud/aws/reconcile.go:91]: we should be using here
```
	RecommendedLeaseDuration     = "137s"
	RecommendedRenewDeadline     = "107s"
	RecommendedRetryPeriod       = "26s"`
```
from `support/config`

File: control-plane-operator/controllers/hostedcontrolplane/kcm/params.go (hunk 1 above)
Comment (enxebre) [reply control-plane-operator/controllers/hostedcontrolplane/kcm/params.go:107]: I think this should be fine as it's, i.e. just omitting the cloud provider flag.
Let's believe these docs for now https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager and be consistent with what we've been doing for kubevirt.

```
kube-apiserver and kube-controller-manager MUST NOT specify the --cloud-provider flag. This ensures that it does not run any cloud specific loops that would be run by cloud controller manager. In the future, this flag will be deprecated and removed.
kubelet must run with --cloud-provider=external. This is to ensure that the kubelet is aware that it must be initialized by the cloud controller manager before it is scheduled any work.
```

**Comment by enxebre** [comment]
ControlPlane Upgrade failing because payload rollout doesn't go beyond partial because can't be schedule because of the new taint https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestUpgradeControlPlane_PreTeardownClusterDump/hostedcluster-example-7pxcw/namespaces/openshift-console-operator/pods/console-operator-5866cdb8f8-zr4wl/console-operator-5866cdb8f8-zr4wl.yaml

[NodePool InPlace upgrade:](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/hostedcluster-example-2gb5k/cluster-scoped-resources/core/nodes/ip-10-0-143-62.ec2.internal.yaml)
```
  conditions:
  - lastHeartbeatTime: "2023-03-10T21:23:01Z"
    lastTransitionTime: "2023-03-10T21:24:32Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: MemoryPressure
  - lastHeartbeatTime: "2023-03-10T21:23:01Z"
    lastTransitionTime: "2023-03-10T21:24:32Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: DiskPressure
  - lastHeartbeatTime: "2023-03-10T21:23:01Z"
    lastTransitionTime: "2023-03-10T21:24:32Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: PIDPressure
  - lastHeartbeatTime: "2023-03-10T21:23:01Z"
    lastTransitionTime: "2023-03-10T21:24:32Z"
    message: Kubelet stopped posting node status.
    reason: NodeStatusUnknown
    status: Unknown
    type: Ready
```
Machine config daemon pod seems to be scheduled https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/hostedcluster-example-2gb5k/namespaces/example-2gb5k-us-east-1c-upgrade/pods/machine-config-daemon-ip-10-0-143-62.ec2.internal/machine-config-daemon-ip-10-0-143-62.ec2.internal.yaml

But logs are empty https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/hostedcluster-example-2gb5k/namespaces/example-2gb5k-us-east-1c-upgrade/pods/machine-config-daemon-ip-10-0-143-62.ec2.internal/machine-config-daemon/machine-config-daemon/logs/

In place upgrader seems fine https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/namespaces/e2e-clusters-dfl2q-example-2gb5k/core/pods/logs/hosted-cluster-config-operator-545cb94d4-sc9wq-hosted-cluster-config-operator.log

@yuqi-zhang does this ring any bell? Any change in 4.14 that might be preventing the machine config pod from doing their job?

**Comment by enxebre** [comment]
/retitle HOSTEDCP-919: Add AWS cloud controller manager

**Comment by yuqi-zhang** [comment]
Hmm, so, just from the logs you posted, I think what happened was that the inplace upgrade started, the node drained, presumably the node got to the reboot, and then never came back up again. I do see pods being drained in the logs, corresponding to the node the MCD was scheduled on `ip-10-0-143-62.ec2.internal`. The fact that the MCD logs are empty probably means it is currently down, meaning the node probably rebooted at some point. https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/hostedcluster-example-2gb5k/cluster-scoped-resources/core/nodes/ip-10-0-143-62.ec2.internal.yaml Seems to indicate the MCD ack'ed the finished drain and then rebooted, so this all lines up


In master we've had a lot of changes recently for the rhel9.2 bump https://github.com/openshift/machine-config-operator/commits/master so it is hard to say. I'm leaning towards the actual node never came back up as opposed to any MCD issue. Are you able to access the console for the node?

**Comment by yuqi-zhang** [comment]
oh wait they're there: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/machine-console-logs/example-2gb5k-us-east-1c-mwmrs.log

**Comment by yuqi-zhang** [comment]
Assuming this is from https://github.com/openshift/machine-config-operator/blob/4f48e1737ffc01b3eb991f22154fc3696da53737/templates/common/_base/files/configure-ovs-network.yaml it doesn't seem to have been changed that recently, so maybe it has something to do with rhel9?

**Comment by yuqi-zhang** [comment]
It looks like since the node never came up we were unable to get the node journals:
https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2271/pull-ci-openshift-hypershift-main-e2e-aws/1634291913915895808/artifacts/e2e-aws/run-e2e/artifacts/TestInPlaceUpgradeNodePool_PreTeardownClusterDump/machine-journals/

Do you have any hints on why DHCP appears to have failed there?

**Comment by csrwng** [comment]
@yuqi-zhang I created https://issues.redhat.com/browse/OCPBUGS-10218 to track the issue. I captured both the log of the mcd and the console of the instance while interactively running the test. Originally assigned to the MCO team but feel free to reassign to the appropriate team. It does sound like it could be a RHEL 9 issue.

@enxebre I'm going to skip both of the in-place upgrade tests for now so we can get unstuck. When OCPBUGS-10218 is fixed, we can re-enable. If the only test that fails is the control plane upgrade test, we should be able to merge, given that other PRs should pass once we get a CI build accepted.

File: api/v1beta1/hostedcluster_types.go [hunk 3]
```diff
@@ -1110,6 +1110,10 @@ type AWSRolesRef struct {
 	//  "Statement": [
 	//    {
 	//      "Action": [
+	//        "autoscaling:DescribeAutoScalingGroups",
```
Comment (enxebre) [inline api/v1beta1/hostedcluster_types.go:1113]: Just for the record, we don't use autoscaling groups, so this should be transparent for us.

File: control-plane-operator/controllers/hostedcontrolplane/cloud/aws/manifests.go [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+func CCMServiceAccount(ns string) *corev1.ServiceAccount {
+	return &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "aws-cloud-controller-manager",
+			Namespace: ns,
+		},
+	}
+}
+
+func CCMRole(ns string) *rbacv1.Role {
```
Comment (enxebre) [inline control-plane-operator/controllers/hostedcontrolplane/cloud/aws/manifests.go:19]: I don't see where we are using `CCMRoleBinding` and `CCMRole`?

**Comment by enxebre** [comment]
We are skipping in place NodePools tests to get CI unblocked. We'll troubleshoot here https://issues.redhat.com/browse/OCPBUGS-10218.
Upgrade test won't pass given our "previous" release points to 4.14 atm https://github.com/openshift/release/pull/37188.
/override ci/prow/e2e-aws

File: api/v1beta1/hostedcluster_types.go (hunk 3 above)
Comment (csrwng) [reply api/v1beta1/hostedcluster_types.go:1113]: Agree, this was added to reflect the list of permissions in AWS documentation: 
https://cloud-provider-aws.sigs.k8s.io/prerequisites/#iam-policies

**Pull request merged**